### ISSUE TITLE
[no-undef] `this` in `typeof this` is reported as undefined incorrectly

### ISSUE THREAD

**Issue by Zzzen** [typescript-eslint/typescript-eslint#4364]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "no-undef": []
  }
}
```

```TS
// Add code below to valid cases of `packages/eslint-plugin/tests/eslint-rules/no-undef.test.ts`
// and run `yarn test no-undef`
const obj = {
  foo: '',
  bar() {
    let foo: typeof this.foo;
  },
};
```

**Expected Result**

No error should be reported.

**Actual Result**

`this` in `typeof this` is reported as undefined.

```
  ● no-undef › valid › 
const obj = {
  foo: '',
  bar() {
    let foo: typeof this.foo;
  },
};
    

    assert.strictEqual(received, expected)

    Expected value to strictly be equal to:
      0
    Received:
      1
    
    Message:
      Should have no errors but had 1: [
      {
        ruleId: 'no-undef',
        severity: 1,
        message: "'this' is not defined.",
        line: 5,
        column: 21,
        nodeType: 'Identifier',
        messageId: 'undef',
        endLine: 5,
        endColumn: 25
      }
    ]
```

**Additional Info**

**Versions**

Tested on latest commit, a9eb0b9e.

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `X.Y.Z` |
| `@typescript-eslint/parser`        | `X.Y.Z` |
| `TypeScript`                       | `X.Y.Z` |
| `ESLint`                           | `X.Y.Z` |
| `node`                             | `X.Y.Z` |

**Comment by bradzacher** [comment]
[repro](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQIPYFoCuCAmEAZqgFxoQBOlGlqYIwAsAFADGGCUKGARgFZgAvGADerMGCIYM5AORyANBLC8AhpQAUASjErJMCCmmywSAJ4AHCBiJmAFnCgA6EwG4VAX2UtPboA&rules=N4XyA&tsConfig=N4XyA)
